spec/models/academic_year_spec.rb

```diff
@@ -12,18 +12,54 @@
 require "rails_helper"
 
 RSpec.describe AcademicYear, type: :model do
-  subject(:academic_year) do
-    build(:academic_year,
-          name: "2023 to 2024",
-          starts_on: Date.current,
-          ends_on: Date.current + 1.day)
-  end
+  describe "with validations" do
+    subject(:academic_year) do
+      build(:academic_year,
+            name: "2023 to 2024",
+            starts_on: Date.current,
+            ends_on: Date.current + 1.day)
+    end
 
-  context "with validations" do
     it { is_expected.to validate_presence_of(:name) }
     it { is_expected.to validate_presence_of(:starts_on) }
     it { is_expected.to validate_presence_of(:ends_on) }
-
     it { is_expected.to validate_comparison_of(:ends_on).is_greater_than_or_equal_to(:starts_on) }
   end
+
+  describe ".for_date" do
+    let!(:existing_academic_year) do
+      create(:academic_year,
+             name: "2024 to 2025",
+             starts_on: Date.new(2024, 9, 1),
+             ends_on: Date.new(2025, 8, 31))
+    end
+
+    context "when date is within an existing academic year" do
+      it "returns the existing academic year" do
+        date = Date.new(2024, 11, 7)
+        expect(described_class.for_date(date)).to eq(existing_academic_year)
+      end
+    end
+
+    context "when date is not within an existing academic year" do
+      it "creates a new academic year" do
+        date = Date.new(2026, 10, 5)
+        expect { described_class.for_date(date) }.to change(described_class, :count).by(1)
+        new_academic_year = described_class.for_date(date)
+        expect(new_academic_year.starts_on).to eq(Date.new(2026, 9, 1))
```

Comment:
```suggestion
        expect(new_academic_year.starts_on).to eq(Date.parse("1 September 2026"))
```
Using strings instead of numbers improves readability